action/evmtx_container.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2024 IoTeX Foundation
+// This source code is provided 'as is' and no warranties are given as to title or non-infringement, merchantability
+// or fitness for purpose and, to the extent permitted by law, all liability for your use of the code is disclaimed.
+// This source code is governed by Apache License 2.0 that can be found in the LICENSE file.
+
+package action
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+	"github.com/pkg/errors"
+)
+
+var (
+	_ EvmTxContainer = (*evmTxContainer)(nil)
+)
+
+type evmTxContainer struct {
+	raw []byte
+	tx  *types.Transaction
+}
+
+func (etx *evmTxContainer) hash() hash.Hash256 {
+	h := etx.tx.Hash()
+	return hash.BytesToHash256(h[:])
+}
+
+func (etx *evmTxContainer) typeToEncoding() (iotextypes.Encoding, error) {
+	switch etx.tx.Type() {
+	case types.LegacyTxType:
+		if etx.tx.Protected() {
+			return iotextypes.Encoding_ETHEREUM_EIP155, nil
+		} else {
+			// tx has pre-EIP155 signature
+			return iotextypes.Encoding_ETHEREUM_UNPROTECTED, nil
+		}
+	default:
+		return 0, ErrNotSupported
+	}
+}
+
+func (etx *evmTxContainer) proto() *iotextypes.EvmTxContainer {
+	if len(etx.raw) == 0 {
+		var err error
+		etx.raw, err = etx.tx.MarshalBinary()
+		if err != nil {
+			panic(err.Error())
+		}
+	}
+	return &iotextypes.EvmTxContainer{
+		Raw: etx.raw,
+	}
+}
+
+func (etx *evmTxContainer) loadProto(pbAct *iotextypes.EvmTxContainer) error {
+	if pbAct == nil {
+		return ErrNilProto
+	}
+	if etx == nil {
+		return ErrNilAction
+	}
+	var (
+		raw = pbAct.GetRaw()
+		tx  = types.Transaction{}
+	)
+	if err := tx.UnmarshalBinary(raw); err != nil {
+		return err
+	}
+	etx.raw = make([]byte, len(raw))
+	copy(etx.raw, raw)
+	etx.tx = &tx
+	return nil
+}
+
+func (etx *evmTxContainer) Unfold(selp *SealedEnvelope, ctx context.Context, checker func(context.Context, *common.Address) (bool, bool, bool, error)) error {
+	var (
+		elp        Envelope
+		elpBuilder = (&EnvelopeBuilder{}).SetChainID(selp.ChainID())
+	)
+	isContract, isStaking, isRewarding, err := checker(ctx, etx.tx.To())
+	if err != nil {
+		return err
+	}
+	if isContract {
+		elp, err = elpBuilder.BuildExecution(etx.tx)
+	} else if isStaking {
+		elp, err = elpBuilder.BuildStakingAction(etx.tx)
+	} else if isRewarding {
+		elp, err = elpBuilder.BuildRewardingAction(etx.tx)
+	} else {
+		elp, err = elpBuilder.BuildTransfer(etx.tx)
+	}
+	if err != nil {
+		return err
+	}
+	encoding, err := etx.typeToEncoding()
+	if err != nil {
+		return err
+	}
+	selp.Envelope = elp
+	selp.encoding = encoding
+	selp.hash = hash.ZeroHash256
+	selp.srcAddress = nil
+	return nil
+}
+
+func (etx *evmTxContainer) Cost() (*big.Int, error) {
```

Comment:
Is the `Cost` method of other specific action no longer being used?

Reply:
can comment that it's only used for actpool, actual cost is based on the underlying action